File: reference/go/internal/jrxmlutil/prompt.go
package jrxmlutil

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"unsafe"
)

// Choice is the user's answer to an interactive prompt.
type Choice int

const (
	Yes Choice = iota
	No
	All
	Skip
)

type option struct {
	label  string
	choice Choice
}

var options = []option{
	{"Yes", Yes},
	{"No", No},
	{"All", All},
	{"Skip file", Skip},
}

// promptFile is the file used for interactive input: os.Stdin when it is a
// real terminal, otherwise the controlling terminal (/dev/tty) when one is
// available — git runs hooks with /dev/null as stdin, so without this fallback
// interactive prompts under `git commit` could never read input. Piped input
// (a pipe or regular file) is respected as-is. promptFile stays nil when
// promptOverride is set (SetPromptInput), which forces line-based prompts.
var promptFile *os.File

// promptReader is shared across all line-based prompts so that buffered input
// is not lost between prompts.
var promptReader *bufio.Reader

// promptOverride, when set via SetPromptInput, replaces the real prompt input
// source. It exists so tests can feed canned yes/no/all/skip answers through
// the real line-based prompt path; it is nil in normal operation.
var promptOverride io.Reader

// SetPromptInput replaces the input source used by line-based prompts with r.
// It is intended for tests that drive interactive prompts with canned answers;
// normal callers never need it. Setting it disables the terminal (raw-mode)
// prompt path, so the arrow-key UI is not used while an override is active.
func SetPromptInput(r io.Reader) {
	promptOverride = r
	promptFile = nil
	promptReader = nil
}

func initPromptInput() {
	if promptReader != nil {
		return
	}
	if promptOverride != nil {
		promptReader = bufio.NewReader(promptOverride)
		return
	}
	if promptFile != nil {
		return
	}
	promptFile = os.Stdin
	if info, err := os.Stdin.Stat(); err == nil && shouldFallbackToTTY(info, isRealTerminal(os.Stdin)) {
		// char device but not a TTY (e.g. /dev/null under git hooks):
		// fall back to the controlling terminal.
		if tty, err := os.OpenFile("/dev/tty", os.O_RDWR, 0); err == nil && isRealTerminal(tty) {
			promptFile = tty
		}
	}
	promptReader = bufio.NewReader(promptFile)
}

// shouldFallbackToTTY reports whether stdin is a character device that is not
// a real terminal (e.g. /dev/null under git hooks) — the case where the prompt
// should try the controlling terminal (/dev/tty) for interactive input. Piped
// input (a pipe or regular file) is respected as-is.
func shouldFallbackToTTY(info os.FileInfo, stdinIsTTY bool) bool {
	return info.Mode()&os.ModeCharDevice != 0 && !stdinIsTTY
}

// Prompt asks the user a yes/no/all/skip question (question should end with
// "?"). In a terminal it renders arrow-key selectable options; otherwise it
// falls back to line-based y/n/a/s input. The default answer is No.
func Prompt(question string) Choice {
	initPromptInput()
	if promptFile == nil || !isRealTerminal(promptFile) {
		return promptLine(question)
	}

	restore, err := enableRawMode(promptFile)
	if err != nil {
		return promptLine(question)
	}
	defer restore()

	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
	defer signal.Stop(sig)
	go func() {
		<-sig
		restore()
		fmt.Fprintln(os.Stderr)
		os.Exit(130)
	}()

	sel := 1 // default to "No"
	for {
		renderPrompt(question, sel)
		key, err := readKeyRaw(promptFile)
		if err != nil {
			fmt.Println()
			return No
		}
		switch key {
		case "left", "up":
			sel--
			if sel < 0 {
				sel = len(options) - 1
			}
		case "right", "down":
			sel++
			if sel >= len(options) {
				sel = 0
			}
		case "enter":
			fmt.Println()
			return options[sel].choice
		case "y", "Y":
			fmt.Println()
			return Yes
		case "n", "N":
			fmt.Println()
			return No
		case "a", "A":
			fmt.Println()
			return All
		case "s", "S":
			fmt.Println()
			return Skip
		}
	}
}

// PromptText asks a free-form question and returns the user's answer as a
// single trimmed line, or "" when the input ends or is blank. It always uses
// line-based input (free-form text needs a normal, canonical terminal — the
// arrow-key raw-mode UI is only for the fixed-choice Prompt).
func PromptText(question string) string {
	initPromptInput()
	fmt.Printf("  %s ", question)
	line, err := promptReader.ReadString('\n')
	if err != nil && line == "" {
		fmt.Println()
		return ""
	}
	return strings.TrimSpace(line)
}

func promptLine(question string) Choice {
	fmt.Printf("  %s [y]es/[n]o/[a]ll/[s]kip (default no): ", question)
	line, err := promptReader.ReadString('\n')
	if err != nil && line == "" {
		fmt.Println("no")
		return No
	}
	switch strings.ToLower(strings.TrimSpace(line)) {
	case "y", "yes":
		return Yes
	case "a", "all":
		return All
	case "s", "skip":
		return Skip
	default:
		return No
	}
}

func renderPrompt(question string, sel int) {
	fmt.Printf("\r\x1b[2K%s  ", question)
	for i, o := range options {
		if i == sel {
			fmt.Printf("\x1b[7m%s\x1b[0m  ", o.label)
		} else {
			fmt.Printf("%s  ", o.label)
		}
	}
}

// isRealTerminal reports whether f is an actual terminal (TCGETS succeeds).
// A plain character device such as /dev/null is not a terminal.
func isRealTerminal(f *os.File) bool {
	fd := f.Fd()
	var t syscall.Termios
	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(syscall.TCGETS), uintptr(unsafe.Pointer(&t)))
	return errno == 0
}

// readKeyRaw reads a single logical key from a terminal already in raw mode.
func readKeyRaw(f *os.File) (string, error) {
	var b [3]byte
	n, err := f.Read(b[:1])
	if err != nil {
		return "", err
	}
	if n == 0 {
		return "", io.EOF
	}
	switch b[0] {
	case 0x1b: // ESC, likely an arrow sequence
		n, err = f.Read(b[1:3])
		if err != nil || n < 2 || b[1] != '[' {
			return "esc", nil
		}
		switch b[2] {
		case 'C':
			return "right", nil
		case 'D':
			return "left", nil
		case 'A':
			return "up", nil
		case 'B':
			return "down", nil
		}
		return "esc", nil
	case '\r', '\n':
		return "enter", nil
	default:
		return string(b[0]), nil
	}
}

// enableRawMode switches the terminal into non-canonical, no-echo mode and
// returns a function that restores the previous settings.
func enableRawMode(f *os.File) (restore func(), err error) {
	fd := f.Fd()
	var old syscall.Termios
	if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(syscall.TCGETS), uintptr(unsafe.Pointer(&old))); errno != 0 {
		return func() {}, errno
	}
	raw := old
	raw.Lflag &^= syscall.ICANON | syscall.ECHO
	raw.Iflag &^= syscall.ICRNL | syscall.IXON
	raw.Cc[syscall.VMIN] = 1
	raw.Cc[syscall.VTIME] = 0
	if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(syscall.TCSETS), uintptr(unsafe.Pointer(&raw))); errno != 0 {
		return func() {}, errno
	}
	return func() {
		_, _, _ = syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(syscall.TCSETS), uintptr(unsafe.Pointer(&old)))
	}, nil
}

File: src/main/java/com/myhooks/diffui/Prompt.java
package com.myhooks.diffui;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.io.Reader;

/**
 * Interactive yes/no/all/skip prompt. On a real terminal it renders
 * arrow-key-selectable options via the controlling terminal in raw mode
 * (default answer No); on a non-terminal it falls back to line-based
 * {@code y/n/a/s} input. The controlling terminal ({@code /dev/tty}) is used
 * rather than stdin because git runs hooks with stdin bound to {@code /dev/null}.
 */
public final class Prompt {

    private Prompt() {
    }

    public static Choice ask(String question) {
        Tty tty = Tty.openRaw();
        if (tty != null) {
            try {
                return askRaw(question, tty);
            } catch (Exception ignored) {
                // fall through to line-based input
            } finally {
                tty.close();
            }
        }
        return askLine(question, new BufferedReader(new InputStreamReader(System.in)), System.out);
    }

    /** Line-based prompt over an injected reader (used by tests). */
    public static Choice ask(String question, Reader in) {
        return askLine(question, toBufferedReader(in), System.out);
    }

    // ------------------------------------------------------------------
    // Line-based fallback
    // ------------------------------------------------------------------

    static Choice askLine(String question, BufferedReader reader, PrintStream out) {
        out.print("  " + question + " [y]es/[n]o/[a]ll/[s]kip (default no): ");
        out.flush();
        String line = readLine(reader);
        if (line == null) {
            out.println("no");
            return Choice.NO;
        }
        return parseLineAnswer(line);
    }

    static Choice parseLineAnswer(String line) {
        switch (line.trim().toLowerCase()) {
            case "y":
            case "yes":
                return Choice.YES;
            case "a":
            case "all":
                return Choice.ALL;
            case "s":
            case "skip":
                return Choice.SKIP;
            default:
                return Choice.NO;
        }
    }

    // ------------------------------------------------------------------
    // Raw terminal mode
    // ------------------------------------------------------------------

    static Choice askRaw(String question, Tty tty) {
        PrintStream out = tty.out();
        int selected = 1; // default answer is No
        while (true) {
            out.print(renderPrompt(question, selected));
            out.flush();
            String key = readKey(tty);
            switch (key) {
                case "left":
                case "up":
                    selected = selected - 1 < 0 ? Choice.values().length - 1 : selected - 1;
                    break;
                case "right":
                case "down":
                    selected = (selected + 1) % Choice.values().length;
                    break;
                case "enter":
                    newline(out);
                    drain(tty);
                    return Choice.values()[selected];
                case "y":
                case "Y":
                    newline(out);
                    drain(tty);
                    return Choice.YES;
                case "n":
                case "N":
                    newline(out);
                    drain(tty);
                    return Choice.NO;
                case "a":
                case "A":
                    newline(out);
                    drain(tty);
                    return Choice.ALL;
                case "s":
                case "S":
                    newline(out);
                    drain(tty);
                    return Choice.SKIP;
                case "":
                    return Choice.NO; // EOF (terminal gone)
                default:
                    break; // ignore ESC / unknown keys
            }
        }
    }

    private static void newline(PrintStream out) {
        out.println();
        out.flush();
    }

    /** Discards typeahead (the rest of "yes"/"no"/"all"/"skip" and the newline). */
    private static void drain(KeySource in) {
        try {
            while (in.readTimed() >= 0) {
                // discard
            }
        } catch (IOException ignored) {
            // best-effort
        }
    }

    /**
     * Reads one logical key. Returns "" on EOF or read error. A real terminal
     * distinguishes a lone ESC from an arrow key with a short timed read; a
     * plain {@link Reader} (tests) reads the escape sequence directly.
     */
    static String readKey(KeySource in) {
        try {
            int c = in.read();
            if (c < 0) {
                return "";
            }
            if (c == 0x1b) {
                int c1 = in.readTimed();
                if (c1 != '[') {
                    return "esc";
                }
                int c2 = in.readTimed();
                return switch (c2) {
                    case 'C' -> "right";
                    case 'D' -> "left";
                    case 'A' -> "up";
                    case 'B' -> "down";
                    default -> "esc";
                };
            }
            if (c == '\r' || c == '\n') {
                return "enter";
            }
            return String.valueOf((char) c);
        } catch (IOException e) {
            return "";
        }
    }

    /** Renders the option row, inverting the selected option. */
    static String renderPrompt(String question, int selected) {
        StringBuilder sb = new StringBuilder();
        sb.append("\r\u001b[2K").append(question).append("  ");
        Choice[] options = Choice.values();
        for (int i = 0; i < options.length; i++) {
            if (i == selected) {
                sb.append("\u001b[7m").append(options[i].label()).append("\u001b[0m  ");
            } else {
                sb.append(options[i].label()).append("  ");
            }
        }
        return sb.toString();
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static BufferedReader toBufferedReader(Reader in) {
        return in instanceof BufferedReader buffered ? buffered : new BufferedReader(in);
    }

    private static String readLine(BufferedReader reader) {
        try {
            return reader.readLine();
        } catch (IOException e) {
            return null;
        }
    }
}
